overlord/snapstate/autorefresh_gating_test.go

```diff
@@ -0,0 +1,478 @@
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/snapstate/snapstatetest"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/testutil"
+
+	. "gopkg.in/check.v1"
+)
+
+type autorefreshGatingSuite struct {
+	testutil.BaseTest
+	state *state.State
+	repo  *interfaces.Repository
+}
+
+var _ = Suite(&autorefreshGatingSuite{})
+
+func (s *autorefreshGatingSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+	dirs.SetRootDir(c.MkDir())
+	s.BaseTest.AddCleanup(func() {
+		dirs.SetRootDir("/")
+	})
+	s.state = state.New(nil)
+
+	s.repo = interfaces.NewRepository()
+	for _, iface := range builtin.Interfaces() {
+		c.Assert(s.repo.AddInterface(iface), IsNil)
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	ifacerepo.Replace(s.state, s.repo)
+}
+
+func (s *autorefreshGatingSuite) mockInstalledSnap(c *C, snapYaml string, hasHook bool) *snap.Info {
+	snapInfo := snaptest.MockSnap(c, string(snapYaml), &snap.SideInfo{
+		Revision: snap.R(1),
+	})
+
+	snapName := snapInfo.SnapName()
+	si := &snap.SideInfo{RealName: snapName, SnapID: "id", Revision: snap.R(1)}
+	snapstate.Set(s.state, snapName, &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+		Current:  si.Revision,
+		SnapType: string(snapInfo.Type()),
+	})
+
+	if hasHook {
+		c.Assert(os.MkdirAll(snapInfo.HooksDir(), 0775), IsNil)
+		err := ioutil.WriteFile(filepath.Join(snapInfo.HooksDir(), "gate-auto-refresh"), nil, 0755)
+		c.Assert(err, IsNil)
+	}
+	return snapInfo
+}
+
+const baseSnapAyaml = `name: base-snap-a
+type: base
+`
+
+const snapAyaml = `name: snap-a
+type: app
+base: base-snap-a
+`
+
+const baseSnapByaml = `name: base-snap-b
+type: base
+`
+
+const snapByaml = `name: snap-b
+type: app
+base: base-snap-b
+version: 1
+`
+
+const kernelYaml = `name: kernel
+type: kernel
+version: 1
+`
+
+const gadget1Yaml = `name: gadget
+type: gadget
+version: 1
+`
+
+const snapCyaml = `name: snap-c
+type: app
+version: 1
+`
+
+const snapDyaml = `name: snap-d
+type: app
+version: 1
+slots:
+    slot: desktop
+`
+
+const snapEyaml = `name: snap-e
+type: app
+version: 1
+base: other-base
+plugs:
+    plug: desktop
+`
+
+const snapFyaml = `name: snap-f
+type: app
+version: 1
+plugs:
+    plug: desktop
+`
+
+const snapGyaml = `name: snap-g
+type: app
+version: 1
+base: other-base
+plugs:
+    desktop:
+    mir:
+`
+
+const coreYaml = `name: core
+type: os
+version: 1
+slots:
+    desktop:
+    mir:
+`
+
+const core18Yaml = `name: core18
+type: os
+version: 1
+`
+
+const snapdYaml = `name: snapd
+version: 1
+type: snapd
+slots:
+    desktop:
+`
+
+func (s *autorefreshGatingSuite) TestAffectedByBase(c *C) {
+	restore := release.MockOnClassic(true)
+	defer restore()
+
+	st := s.state
+
+	st.Lock()
+	defer st.Unlock()
+	s.mockInstalledSnap(c, snapAyaml, true)
+	baseSnapA := s.mockInstalledSnap(c, baseSnapAyaml, false)
```

Comment:
(nitpick) not a fan of the bool args in the signaureHere, maybe use `const useHook = true` and `const noHook = false` and then pass this for clarity?

Reply:
Nice idea!